### DISCUSSION TITLE
Elasticsearch embedded store search failed in elastic client

### DISCUSSION THREAD

**Discussion by sashaepstein** [langchain4j/langchain4j#2154] [Q&A] [answered]
Greeting! I am trying to use ElasticsearchEmbeddingStore, I can successfully create index and ingest lots of embeddings documents there, but the search fails. Debugging through the elastic client code I can see the server returning 400 without any useful message. I can manually execute the same vector query the elastic client is sending, using curl and content type application/json (the elastic client forces content type: application/vnd.elasticsearch+json; compatible-with=8), and curl successfully returns all found matches. I use the latest version of langchain4j (0.36) that uses elastic client v. 8.15.3, and I tried both, matching elasticsearch server (v. 8.15.3) and v. 8.15.0 as it is shown in your example, with the same result.
Any idea where I should look?
Thanks.

**Comment by sashaepstein** [comment]
Here is the index that was created, based on the metadata structure of my Document:
{
  "rag" : {
    "aliases" : { },
    "mappings" : {
      "properties" : {
        "metadata" : {
          "properties" : {
            "externalId" : {
              "type" : "text",
              "fields" : {
                "keyword" : {
                  "type" : "keyword",
                  "ignore_above" : 256
                }
              }
            },
            "hcpCountry" : {
              "type" : "text",
              "fields" : {
                "keyword" : {
                  "type" : "keyword",
                  "ignore_above" : 256
                }
              }
            },
            "hcpId" : {
              "type" : "text",
              "fields" : {
                "keyword" : {
                  "type" : "keyword",
                  "ignore_above" : 256
                }
              }
            },
            "hcpName" : {
              "type" : "text",
              "fields" : {
                "keyword" : {
                  "type" : "keyword",
                  "ignore_above" : 256
                }
              }
            },
            "index" : {
              "type" : "text",
              "fields" : {
                "keyword" : {
                  "type" : "keyword",
                  "ignore_above" : 256
                }
              }
            },
            "postDate" : {
              "type" : "date"
            },
            "source" : {
              "type" : "text",
              "fields" : {
                "keyword" : {
                  "type" : "keyword",
                  "ignore_above" : 256
                }
              }
            },
            "tagids" : {
              "type" : "text",
              "fields" : {
                "keyword" : {
                  "type" : "keyword",
                  "ignore_above" : 256
                }
              }
            }
          }
        },
        "text" : {
          "type" : "text",
          "fields" : {
            "keyword" : {
              "type" : "keyword",
              "ignore_above" : 256
            }
          }
        },
        "vector" : {
          "type" : "dense_vector",
          "dims" : 384,
          "index" : true,
          "similarity" : "cosine",
          "index_options" : {
            "type" : "int8_hnsw",
            "m" : 16,
            "ef_construction" : 100
          }
        }
      }
    },
    "settings" : {
      "index" : {
        "routing" : {
          "allocation" : {
            "include" : {
              "_tier_preference" : "data_content"
            }
          }
        },
        "number_of_shards" : "1",
        "provided_name" : "rag",
        "creation_date" : "1732098723479",
        "number_of_replicas" : "1",
        "uuid" : "Il-2f8zkSGC9SfBYVGVzLA",
        "version" : {
          "created" : "8512000"
        }
      }
    }
  }
}

**Comment by dadoonet** [comment]
@sashaepstein I'll be happy to help to diagnose that. Is there a sample (Java) code you could share? Like a project repo using maven for example where you connect to a cluster (ideally with TestContainers) and then index some docs and search for them?

**Comment by sashaepstein** [comment]
Unfortunately I can't share the entire code as it is my company proprietary code. But I think I can just put here important excerpts that differ from your sample, and I hope that if you use these parts in your sample you will be able to reproduce the problem.
Some notes about my code:
1. It was originally created with using langchain4j InMemoryEmbeddingStore, and it works fine with that store.
2. I use dev.langchain4j.data.document.Document as the stored data, rather than its component TextSegment as in your example.
3. embeddingStore is now ElasticsearchEmbeddingStore instance created like in your example.
4. embeddingModel is AllMiniLmL6V2EmbeddingModel as in your example
5. Tweet object has "content" String field that is used to create TextSegment and Document.
6. I put Metadata with 7 members in every Document object, as you can see below

Here is the ingesting part (also as I mentioned, ingest works perfectly fine, I can use curl with query_vector query to search vector matches successfully):

        private fun ingestPosts(posts: List<Tweet>): Int {
            val addedPostIds = mutableListOf<String>()
            val documents = posts.filter { it.addedToRag == null }.map {
                if (it.content.isNullOrBlank())
                    null
                else {
                    it.externalId?.let { addedPostIds.add(it) }
                    Document.from(it.content, Metadata().apply {
                        this.put("externalId", it.externalId ?: "")
                        this.put("source", it.source.name)
                        this.put("hcpName", it.tAccount?.name ?: it.tAccount?.username ?: "")
                        this.put("hcpId", it.tAccount?.id ?: "")
                        this.put("hcpCountry", it.tAccount?.countryCode?.lowercase() ?: "")
                        this.put("postDate", it.externalCreatedAt!!.toLocalDateTime().toLocalDate().toString())
                        this.put("tagids", it.tags!!.map { it.id!! }.joinToString (","))
                    })
                }
            }.filterNotNull()
            val ingestor = EmbeddingStoreIngestor.builder()
                .documentSplitter(StoreSegmentSplitter())
                .textSegmentTransformer(StoreSegmentTransformer())
                .embeddingModel(embeddingModel)
                .embeddingStore(embeddingStore)
                .build()
            val startTime = getCurrentTimestamp().getTime()
            ingestor.ingest(documents)

Here is the search part (note: I don't use filter in EmbeddingSearchRequest):

        val builder = EmbeddingStoreContentRetriever.builder()
            .embeddingStore(embeddingStore)
            .embeddingModel(embeddingModel)
            .dynamicMaxResults(this.dynamicMaxResults)
//            builder.dynamicFilter(dynamicFilter)
            .dynamicMinScore(minScore)
        contentRetriever = builder.build()

**Comment by sashaepstein** [comment]
I just noticed that in ElasticsearchEmbeddingStore you use a very concrete object type TextSegment rather than generic Embedded. I think this is the culprit. I will verify it.

**Comment by sashaepstein** [comment, marked as answer]
No, that was a wrong shot. I found the problem, it was max result number that I used in EmbeddingSearchRequest that was over the max recommended #. Once I limited it to 2000 the search started to work.

**Reply by dadoonet** [reply]
Thanks!

**Reply by sashaepstein** [reply]
The ES returned the message: "Result window is too large, from + size must be less than or equal to: [10000]".  But your documentation clearly mentions the max, so I don't think anything should be done on LC4 side for this matter, it was my fault.
However, I have other suggestions that I had to implement in my project to make it production quality. Here they are:
1. Make the document that is pushed to ES, have explicit ID that the dev can specify (rather than assigning a random GUID). Otherwise, I believe, it is impossible to replace existing document, and can lead to duplication of data in the ES index.
2. Add a generic (or Object) Collection type to Metadata.SUPPORTED_VALUE_TYPES, to support IsIn kinds of filters whose comparisonValues() method returns Collection<?>.  It is common to filter on one of the metadata fields.
3. Current implementation of ElasticsearchMetadataFilterMapper.map() is too rigid, it doesn't support custom filters. I think that the Query should be formed not in that class methods like mapEqual, but in each implementation of the Filter interface that should have public Query query() method. Then developers can implement their own filters with the code in query() method to create the needed ES query.

If you are interested, I can share my implementation.
Thanks.

**Reply by sashaepstein** [reply]
1. replacing is simple when it is done by id. Once the PR that you mentioned, is implemented, the invocation of addAll with the existing id should replace the existing document. This would be a seamless update/upsert that I asked for.
2. Certainly. I looked at the PR and am surprised in the amount of code changes just to add the Collection<?> type. I thought that any suggestion to change the existing code should add only the code necessary for that change. Maybe I am missing something there.
3. Generic Filter extensions, like IsIn. should be store agnostic. However, for ES store we can define a set of ES-specific filters like this one that creates a term_set filter (my project is in Kotlin):

            data class TermsSetFilter(val key: String, val comparisonValues: Collection<Any>, val minMatch: Int) : ESFilter() {
                override fun Test() = false
                
                override fun query(): Query {
                    return ( Query.Builder().bool { b: BoolQuery.Builder ->
                        b.filter { f: Query.Builder ->
                            f.termsSet { t: TermsSetQuery.Builder ->
                                t.field(formatKey(key, comparisonValues))
                                val storedScriptId = StoredScriptId.Builder().id(minMatch.toString()).build()
                                val script = Script.Builder().stored(storedScriptId).build()
                                t.minimumShouldMatchScript(script).terms(comparisonValues.map { it: Any? -> it.toString() })
                            }
                        }
                    }.build() as Query)
                }
            }
instead of constantly modifying (or overriding like I did) ElasticsearchMetadataFilterMapper class.

**Reply by sashaepstein** [reply]
I had to override map() in ElasticsearchMetadataFilterMapper class to be able to create any ES filter of any complexity, that keeps the filter specific code inside the filter class:

public class ElasticsearchMetadataFilterMapperImp extends ElasticsearchMetadataFilterMapper {

    public static class ESFilter implements Filter {
        @Override
        public boolean test(Object object) {
            return false;
        }
        public Query query() {
            return null;
        }
        public static String formatKey(String key, Object comparisonValue) {
            return !(comparisonValue instanceof String) && !(comparisonValue instanceof UUID) ? "metadata." + key : "metadata." + key + ".keyword";
        }

        public static String formatKey(String key, Collection<?> comparisonValues) {
            return formatKey(key, comparisonValues.iterator().next());
        }
    }

    public ElasticsearchMetadataFilterMapperImp() {
        super();
    }

    static Query map(Filter filter) {
        if (filter instanceof CustomFilter) {
            return ((CustomFilter) filter).query();
        } else
            return ElasticsearchMetadataFilterMapper.map(filter);
    }
}

**Reply by sashaepstein** [reply]
> `addAll()` will not guarantee upsert. It will only guarantee add. So I guess you will need to `deleteAll(ids)` and then `addAll(ids, embeddings, segments)` for the time being unless the underlying implementation's add action acts as an upsert.

Upcert works in my override of `addAll()` where, just like in the original `ElasticsearchEmbeddingStore`, I set document id in the BulkRequest, as it should be upsert for the underlying elasticsearch.client library.
This is my third project where I use Elastic, and with all 3 underlying libraries that I used: com.sksamuel.elastic4s (Scala), com.jillesvangurp:search-client (kotlin) and directly elasticsearch.client (Java) I always had upcert capabilities specifying document id in the single or bulk operations. That is my, certainly limited, understanding of Elastic :-)

> What parts of the changes do you consider unnecessary there?

It's easier to mention what is needed:
1. update comment on line 63 with the Collection type, not List as it is in the PR, as I see no need to confine the collection type only to List.
2. add Collection.class type to SUPPORTED_VALUE_TYPES
3. In the public Metadata constructor, as this is the single place where the types should be checked, add a check for the collection values, as it is in the PS's `validateValueList`.
4. add `getCollection(String key)` and `put(String key, Collection<?> values)` methods.

That's would be the minimal disturbance of the existing class.
 
> Well, `Filter` was not designed to support extensibility in that way, but we can definitely consider it. If you have a solution in mind, feel free to open a PR and tag me and @dadoonet. Thank you!

I shared my solution in the code excerpts in the above messages. It works well in my case, allowing encapsulation of functionality in the derived filter classes without the need of constant modification of the filter mapper class with every new ES filter capability.
Thanks.

**Reply by sashaepstein** [reply]
> @dadoonet WDYT about adding support for `ElasticsearchFilter implements Filter` in `langchain4j-elasticsearch` module that will allow to specify any filter with Elasticsearch APIs?

Actually, thinking more and trying to address other requirements of our product (Medical.Watch), I came to a conclusion that just specifying query filter (with whatever complex structure) is not enough to fully customize the query. I had to override `internalSearch()` in ConfigurationKnn with the following content:
      `KnnQuery knn = getKnnQuery();`
      `Query myQuery = ....... .query( b -> b.knn(knn)) ...........`
      `return elasticsearchClient.search((sr) -> sr.index(indexName, new String[0]).size(embeddingSearchRequest.maxResults()).query(myQuery).minScore(embeddingSearchRequest.minScore()), Document.class);`
